code/src/ng_link/ng_state.py

```diff
@@ -0,0 +1,348 @@
+from typing import List, Any, Optional, Union
+from .ng_layer import NgLayer
+from pint import UnitRegistry
+from pathlib import Path
+import sys
+import re
+from utils import utils
+
+# IO types
+PathLike = Union[str, Path]
+
+class NgState():
+    def __init__(self, input_config:dict, output_json:PathLike, verbose:Optional[bool]=False) -> None:
+        """
+        Class constructor
+        
+        Parameters
+        ------------------------
+        image_config: dict
+            Dictionary with the image configuration based on neuroglancer documentation.
+        output_json: PathLike
+            Path where the json will be written.
+        
+        verbose: Optional[bool]
+            If true, additional information will be shown. Default False.
+        """
+        
+        self.input_config = input_config
+        self.output_json = Path(self.__fix_output_json_path(output_json))
+        self.verbose = verbose
+        
+        # State and layers attributes
+        self.__state = {}
+        self.dimensions = {}
+        self.__layers = []
+        
+        self.__prefix = 'ng_link_'
+        
+        # Initialize principal attributes
+        self.initialize_attributes(self.input_config)
+    
+    def __fix_output_json_path(self, output_json:PathLike) -> str:
+        
+        """
+        Fixes the json output path in order to have a similar structure for all links.
+        
+        Parameters
+        ------------------------
+        output_json: PathLike
+            Path of the json output path.
+        
+        Returns
+        ------------------------
+        str
+            String with the fixed outputh path.
+        """
+        output_json = Path(
+            str(output_json).replace(

```

Comment:
Is this equivalent to `os.path.basename`?

Reply:
Yeah, I tried using the property tag. However, since I have lists and dictionaries as states Python gets confused if I am actually setting a new list or just retrieving a value. That's why I explicitly used the method (I'll look into that more). 

The links are being saved in the stitched folder: 
![image](https://user-images.githubusercontent.com/36769694/191853086-67d7f1eb-8c8f-40e4-b7d1-51e5badf804a.png)

- ng_link.txt: Contains the actual neuroglancer link to visualize the data.
- ng_link_datasetname.json: Neuroglancer configuration.

**Is this equivalent to os.path.basename?** Not really, the thing is that since we're mounting the bucket in a folder that is created in execution time, neuroglancer needs the actual path where the zarr is saved within the google bucket by `using gs://bucketname/path/to/file.zarr`. So we need to get rid of the path where the bucket is mounted in the VM or container and replace it with the file extension for neuroglancer `zarr://` and the google bucket ext. `gs://`.

Reply:
Got it.  For URI parsing I usually use: https://docs.python.org/3/library/urllib.parse.html

Reply:
`process_output.json`